token_source.go

```diff
@@ -0,0 +1,215 @@
+	if len(items) != 4 {
+		return nil, ErrInvalidClientSecret
+	}
+
+	if !bytes.Equal(items[2], v1SecretTokenIdentifier) {
+		return nil, ErrInvalidClientSecret
+	}
+
+	jwkData, err := base64.RawURLEncoding.DecodeString(string(items[3]))
+	if err != nil {
+		return nil, ErrInvalidClientSecret
+	}
+
+	npk := &jose.JSONWebKey{}
+	err = npk.UnmarshalJSON(jwkData)
+	if err != nil {
+		return nil, ErrInvalidClientSecret
+	}
+
+	if npk.IsPublic() || !npk.Valid() {
+		return nil, ErrInvalidClientSecret
+	}
+
+	_, ok := npk.Key.(ed25519.PrivateKey)
+	if !ok {
+		return nil, ErrInvalidClientSecret
+	}
+
+	return npk, nil
+}
+
+func (c *c1TokenSource) Token() (*oauth2.Token, error) {
+	jsigner, err := jose.NewSigner(
+		jose.SigningKey{
+			Algorithm: jose.EdDSA,
+			Key:       c.clientSecret,
+		},
+		&jose.SignerOptions{
+			NonceSource: &xjwt.RandomNonce{Size: 16},
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	// Our token host may include a port, but the audience never expects a port
+	aud := c.tokenHost
+	host, _, err := net.SplitHostPort(aud)
+	if err == nil {
+		aud = host
+	}
+	now := time.Now()
+	claims := &jwt.Claims{
+		Issuer:    c.clientID,
+		Subject:   c.clientID,
+		Audience:  jwt.Audience{aud},
+		Expiry:    jwt.NewNumericDate(now.Add(time.Minute * 2)),
+		IssuedAt:  jwt.NewNumericDate(now),
+		NotBefore: jwt.NewNumericDate(now.Add(-time.Minute * 2)),
+	}
+	b, err := json.Marshal(claims)
+	if err != nil {
+		return nil, err
+	}
+
+	rv, err := jsigner.Sign(b)
+	if err != nil {
+		return nil, err
+	}
+
+	s, err := rv.CompactSerialize()
+	if err != nil {
+		return nil, err
+	}
+
+	body := url.Values{
+		"client_id":             []string{c.clientID},
+		"grant_type":            []string{"client_credentials"},
+		"client_assertion_type": []string{assertionType},
+		"client_assertion":      []string{s},
+	}
+
+	tokenUrl := url.URL{
+		Scheme: "https",
+		Host:   c.tokenHost,
+		Path:   "auth/v1/token",
+	}
+
+	req, err := http.NewRequestWithContext(c.ctx, http.MethodPost, tokenUrl.String(), strings.NewReader(body.Encode()))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, status.Errorf(codes.Unauthenticated, "failed to get token: %s", resp.Status)
+	}
+
+	c1t := &c1Token{}
+	err = json.NewDecoder(resp.Body).Decode(c1t)
+	if err != nil {
+		return nil, err
+	}
+
+	if c1t.AccessToken == "" {
+		return nil, status.Errorf(codes.Unauthenticated, "failed to get token: empty access token")
+	}
+
+	return &oauth2.Token{
+		AccessToken: c1t.AccessToken,
+		TokenType:   c1t.TokenType,
+		Expiry:      time.Now().Add(time.Duration(c1t.Expiry) * time.Second),
+	}, nil
+}
+
+func NewC1TokenSource(ctx context.Context, clientID string, clientSecret string, hostOverride string) (oauth2.TokenSource, error) {
```

Comment:
naming nit: why does this have C1 in it?

Shouldn't it just be `NewTokenSource`?